### DISCUSSION TITLE
Trying to identify most common distinct motifs in time series data and then search for those motifs

### DISCUSSION THREAD

**Discussion by chrisruk** [TDAmeritrade/stumpy#438] [Help: Coding & Implementations] [answered]
Hi,

I'm wondering anyone can possibly point me in the right direction,

I'm reading the paper - https://www.researchgate.net/publication/343280455_Human_Presence_Detection_by_monitoring_the_indoor_CO2_concentration

They mention:
"We identified events of presence or absence using motif detection on the CO2 concentration time series. Therefore, we used Motif Clustering within computing the full distance matrix using the STOMP algorithm".  

I believe this is one of the algorithms your API uses.

The following is an example of one of their figures:
![fig4](https://user-images.githubusercontent.com/19668907/126041549-f402577e-61ab-4606-90af-d430d1f6ffa3.png)

I have used their open data and tried to use stumpy, to find the same first 3 motifs as them, by doing:

```
#!/usr/bin/python3

import stumpy
import csv
import datetime as dt
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.dates as dates
from matplotlib.patches import Rectangle
import matplotlib.pyplot as plt

if __name__ == "__main__":
    co2 = []
    with open('MUC2020/csv/RoomA_CO2.csv') as csvfile:
        reader = csv.DictReader(csvfile, delimiter=';')
        for row in reader:
            if len(row['CO2-Concentration (in ppm)'].strip()) > 0:
                co2.append(float(row['CO2-Concentration (in ppm)']))
    fig, axs = plt.subplots(1, sharex=True, gridspec_kw={'hspace': 0})
    plt.suptitle('Motif (Pattern) Discovery', fontsize='30')
    axs.plot(co2)
    axs.set_ylabel('CO2 ppm', fontsize='20')
    m = 10
    c = np.array(co2)
    p = stumpy.stump(c, m)
    dists, inde = stumpy.motifs(c, p[:,1], max_matches=3)
    for z in range(0, inde.shape[1]):
        rect = Rectangle((inde[0][z], 0), m, 10, facecolor=(1.0,0.0,0.0))
        axs.add_patch(rect)
    plt.show()
```

The following image shows my results using their data:

![my_fig](https://user-images.githubusercontent.com/19668907/126042029-826b1667-fd83-467a-b416-20812ddd9a98.png)

The 3 motifs that the motifs function is finding for me, all seem to be very similar in that they're all below the downward trends of the CO2 ppm line.

I'm just wondering if anyone could possibly point me in the right direction to find the same 3 unique motifs they have and to then search for those motifs across the time series data.

I'm thinking the motifs function, might not be the function I should be using to find the most common distinct motifs?

Many thanks!

**Comment by seanlaw** [comment, marked as answer]
@chrisruk Thank you for your question and welcome to the STUMPY community. Please be forewarned that the `stumpy.motifs` function is still in the experimental stage and should be officially released in the upcoming v1.9.0 release. In the meantime, the API may still change.

Having said that, it looks like what you've done seems fine. However, I could be wrong but your window size of `m = 10` appears to be smaller than the window size used in the paper. Also, can you please plot the raw CO2 time series along with the matrix profile computed using `stumpy.stump` (like what is shown [here](https://stumpy.readthedocs.io/en/latest/Tutorial_STUMPY_Basics.html#Find-a-Motif-Using-STUMP))? That will at least give you an idea of where the potential motifs are located within your time series (by looking at the major local minima). Of course, the matrix profile will depend on the appropriate window size used.

**Reply by chrisruk** [reply]
Hi Sean, thanks a lot for your reply.

I've generated a matrix profile now, using the same code as before, but increasing the window size to 18.

![Figure_1](https://user-images.githubusercontent.com/19668907/126047455-f5b82a44-b927-4ea3-ad04-c3067f0514e7.png)

The motifs from the motifs function seem to have changed somewhat, but they still seem to be similar to each other.  Rather than  say a motif for the increasing CO2 and another for the decreasing CO2 section.

New code:
```
#!/usr/bin/python3

import stumpy
import csv
import datetime as dt
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.dates as dates
from matplotlib.patches import Rectangle
import matplotlib.pyplot as plt

if __name__ == "__main__":
    co2 = []
    with open('MUC2020/csv/RoomA_CO2.csv') as csvfile:
        reader = csv.DictReader(csvfile, delimiter=';')
        for row in reader:
            if len(row['CO2-Concentration (in ppm)'].strip()) > 0:
                co2.append(float(row['CO2-Concentration (in ppm)']))
    fig, axs = plt.subplots(2, sharex=True, gridspec_kw={'hspace': 0})
    plt.suptitle('Motif (Pattern) Discovery', fontsize='30')
    axs[0].plot(co2)
    axs[0].set_ylabel('CO2 ppm', fontsize='20')
    m = 18

    c = np.array(co2)
    p = stumpy.stump(c, m)

    dists, inde = stumpy.motifs(c, p[:,1], max_matches=3)
    for z in range(0, inde.shape[1]):
        rect = Rectangle((inde[0][z], 0), m, 10, facecolor=(1.0,0.0,0.0))
        axs[0].add_patch(rect)

    axs[1].plot(p[:, 0])
    axs[1].set_ylabel('Matrix profile', fontsize='20')

    plt.show()
```

**Reply by seanlaw** [reply]
Any chance you can upload the `RoomA_CO2.csv` file here (you can drag and drop files directly into a "reply" text box) below?  Also, did you happen to downsample the data? The details in the your time series (i.e., total number of data points) look different from the original published figure.

@mexxexx Do you have any suggestions/ideas here?

**Reply by chrisruk** [reply]
Sure, I've just attached the file - [RoomA_CO2.csv](https://github.com/TDAmeritrade/stumpy/files/6835348/RoomA_CO2.csv)

I got the data from - https://mygit.th-deg.de/tcg/MUC2020/

That's a good question re. downsampling, I didn't downsample the data, but I just noticed it says the following about the data they provide -

"This station allows measuring the CO2 concentration with a resolution up to 10 minutes (the resolution of the published dataset was down sampled to 30 minutes)."

**Reply by seanlaw** [reply]
I'll preface my response with a few points and reiterations:

1. Without more information from the originally published paper (i.e., window size, no downsampling, matrix profile), it'll be very hard to reproduce their motifs without manual fine tuning
2. According to the [`stumpy.motifs` API documentation](https://stumpy.readthedocs.io/en/latest/api.html#stumpy.motifs), this function requires `P`, the matrix profile of your time series `T`, which, in your case should be `c[:, 0]` and not `c[:, 1]`. `c[:, 1]` are the matrix profile indices and not the matrix profile distance values themselves. 
3. The [`stumpy.motifs` API documentation](https://stumpy.readthedocs.io/en/latest/api.html#stumpy.motifs) had an incorrect description for `max_motifs` and that has been corrected now. 
4. The indices (ndarray) that are returned by `stumpy.motifs` will, at most, have `max_motifs` number of rows and `max_matches` number of columns (but possibly less if the constraints are not satisfied). Since your original call to `stumpy.motifs` did not provide `max_motifs` then `max_motifs` defaults to `1`. This means that your original results only returned the top-1 motif along with its `2` nearest-neighbors (outside of the default exclusion zone(s)). The first column is the motif index itself or the "self-match" or "trivial-match". So, if you want `3` nearest-neighbors then you need to set `max_matches=4` (`1` trivial/self-match along with `3` or less nearest neighbors - again, this will depend on whether or not the constraints are satisfied so you may possibly get less!)

Here's what I ended up with:

```
import stumpy
import csv
import numpy as np
import matplotlib.pyplot as plt

plt.rcParams["figure.figsize"] = [20, 6]  # width, height
plt.rcParams['xtick.direction'] = 'out'

co2 = []
with open('Downloads/RoomA_CO2.csv') as csvfile:
    reader = csv.DictReader(csvfile, delimiter=';')
    for row in reader:
        if len(row['CO2-Concentration (in ppm)'].strip()) > 0:
            co2.append(float(row['CO2-Concentration (in ppm)']))
m = 18

c = np.array(co2)
p = stumpy.stump(c, m)

dists, inde = stumpy.motifs(c, p[:, 0], max_matches=4, max_motifs=3)

fig, axs = plt.subplots(2, sharex=True, gridspec_kw={'hspace': 0})
plt.suptitle('Motif (Pattern) Discovery', fontsize='30')
axs[0].plot(co2)
axs[0].set_ylabel('CO2 ppm', fontsize='20')

for z in range(0, inde.shape[0]):
    start = inde[z, 0]
    stop = inde[z, 0] + m
    axs[0].plot(np.arange(start, stop), c[start : stop], c='red')

axs[1].plot(p[:, 0])
axs[1].set_ylabel('Matrix profile', fontsize='20')

plt.show()
```

![download-4](https://user-images.githubusercontent.com/7473521/126063699-7c03acd5-5224-4ca7-a99c-d267f74e974c.png)

Again, it's nearly impossible to recreate the original published work without more information and/or the exact same data. Hopefully, this helps.

**Reply by chrisruk** [reply]
Thanks a lot for your help, it's much appreciated! 

I adapted your changes, to use the 'stumpy.match' function to find all matches for each motif.

I decided to use just 2 motifs as I'm mainly curious about the entering / leaving of a room.

![Figure_1](https://user-images.githubusercontent.com/19668907/126067045-4190b0c9-aadb-427f-9db1-502ad57b2875.png)

I also added the 'max_distance' parameter to stumpy.match to get a couple more matches.

```
import stumpy
import csv
import numpy as np
import matplotlib.pyplot as plt

plt.rcParams["figure.figsize"] = [20, 6]  # width, height
plt.rcParams['xtick.direction'] = 'out'

co2 = []
with open('MUC2020/csv/RoomA_CO2.csv') as csvfile:
    reader = csv.DictReader(csvfile, delimiter=';')
    for row in reader:
        if len(row['CO2-Concentration (in ppm)'].strip()) > 0:
            co2.append(float(row['CO2-Concentration (in ppm)']))

m = 24
c = np.array(co2)
p = stumpy.stump(c, m)

dists, inde = stumpy.motifs(c, p[:, 0], max_motifs=2)

fig, axs = plt.subplots(2, sharex=True, gridspec_kw={'hspace': 0})
plt.suptitle('Motif (Pattern) Discovery', fontsize='30')
axs[0].plot(co2)
axs[0].set_ylabel('CO2 ppm', fontsize='20')

cols = ['red' , 'green', 'blue' ]

for z in range(0, inde.shape[0]):
    col = cols[z]
    start = inde[z, 0]
    stop = inde[z, 0] + m
    matches = stumpy.match(c[start:stop],c, max_distance=2.0) 
    for mt in range(matches.shape[0]):
        s = matches[mt, 1]
        st = s + m
        axs[0].plot(np.arange(s, st), c[s : st], c=col)

axs[1].plot(p[:, 0])
axs[1].set_ylabel('Matrix profile', fontsize='20')
plt.show()
```

**Reply by seanlaw** [reply]
@chrisruk That's wonderful! It looks like you're already turning into a STUMPY pro. I wanted to provide a couple of additional points in case it may be valuable to you (and others reading this):

1. Underneath the hood, `stumpy.motifs` actually calls `stumpy.match` except that it tries to make some educated guesses. Of course, if you want more fine grained control, calling `stumpy.match` yourself will give you the most flexibility and freedom. Your approach is great!
2. Since it looks like this data has motifs of different lengths, I did a little bit more exploring to see if there were "better" window sizes (using a soon-to-be-released function) and it looks like the optimal window sizes is around 45 data points in length and (maybe) around 90 data points in length. I think that the motifs identified in the paper were too short but I guess I'm use to motifs that are "greedy" and try to include as much information as is useful. Hopefully, you'll find this helpful.

Best of luck with your work!

**Reply by chrisruk** [reply]
Thanks again for all your help! I hadn't realised that re. the stumpy.motifs function using stumpy.match.

What's the soon-to-be-released function going to be called out of interest, so I can look out for it.

**Reply by seanlaw** [reply]
> Thanks again for all your help!

You bet! 

> What's the soon-to-be-released function going to be called out of interest, so I can look out for it.

It is called `stumpy.stimp` and actually computes what is referred to as a "pan matrix profile" ([see this paper for more details](https://www.cs.ucr.edu/~eamonn/PAN_SKIMP%20%28Matrix%20Profile%20XX%29.pdf)). A work-in-progress tutorial can be found [here](https://github.com/TDAmeritrade/stumpy/blob/main/docs/Tutorial_Pan_Matrix_Profile.ipynb)

Here is the code that I used for your reference (please be warned that this is still in development and subject to change):

```
min_m, max_m = 10, 100
co2_pan = stumpy.stimp(c, min_m=min_m, max_m=max_m, percentage=1.0)  # This percentage controls the extent of `stumpy.scrump` completion
percent_m = 1.0  # The percentage of windows to compute
n = np.ceil((max_m - min_m) * percent_m).astype(int)
for _ in range(n):
    co2_pan.update()
```

And you can plot the results with:

```
import matplotlib.pyplot as plt
from matplotlib import cm

fig = plt.figure()
fig.canvas.toolbar_visible = False
fig.canvas.header_visible = False
fig.canvas.footer_visible = False

color_map = cm.get_cmap("Greys_r", 256)
threshold = 0.2  # 0.2 is usually an excellent default but this is something that you'll need to play around with
im = plt.imshow(co2_pan.pan(threshold=threshold), cmap=color_map, origin="lower", interpolation="none", aspect="auto")
plt.xlabel("Time", fontsize="20")
plt.ylabel("m", fontsize="20")
plt.clim(0.0, 1.0)
plt.colorbar()
plt.tight_layout()

plt.show()
```

![download-5](https://user-images.githubusercontent.com/7473521/126219501-9f0166a9-a15a-4c27-87bc-74521fa0c58b.png)

As you adjust the `threshold`, what you are looking for is the location of the peak of the right angle triangles that are formed. Of course, this is definitely more art than science but at least it is a "better" way to explore the data.

**Reply by chrisruk** [reply]
Sorry for the delay in replying. I'll have a look at the tutorial and paper you referenced - they look very handy!

**Reply by seanlaw** [reply]
No apologies necessary. Just an FYI that the new version was released today!

**Reply by amurthy-sunysb** [reply]
> Thanks a lot for your help, it's much appreciated!
> 
> I adapted your changes, to use the 'stumpy.match' function to find all matches for each motif.
> 
> I decided to use just 2 motifs as I'm mainly curious about the entering / leaving of a room.
> 
> ![Figure_1](https://user-images.githubusercontent.com/19668907/126067045-4190b0c9-aadb-427f-9db1-502ad57b2875.png)
> 
> I also added the 'max_distance' parameter to stumpy.match to get a couple more matches.
> 
> ```
> import stumpy
> import csv
> import numpy as np
> import matplotlib.pyplot as plt
> 
> plt.rcParams["figure.figsize"] = [20, 6]  # width, height
> plt.rcParams['xtick.direction'] = 'out'
> 
> co2 = []
> with open('MUC2020/csv/RoomA_CO2.csv') as csvfile:
>     reader = csv.DictReader(csvfile, delimiter=';')
>     for row in reader:
>         if len(row['CO2-Concentration (in ppm)'].strip()) > 0:
>             co2.append(float(row['CO2-Concentration (in ppm)']))
> 
> m = 24
> c = np.array(co2)
> p = stumpy.stump(c, m)
> 
> dists, inde = stumpy.motifs(c, p[:, 0], max_motifs=2)
> 
> fig, axs = plt.subplots(2, sharex=True, gridspec_kw={'hspace': 0})
> plt.suptitle('Motif (Pattern) Discovery', fontsize='30')
> axs[0].plot(co2)
> axs[0].set_ylabel('CO2 ppm', fontsize='20')
> 
> cols = ['red' , 'green', 'blue' ]
> 
> for z in range(0, inde.shape[0]):
>     col = cols[z]
>     start = inde[z, 0]
>     stop = inde[z, 0] + m
>     matches = stumpy.match(c[start:stop],c, max_distance=2.0) 
>     for mt in range(matches.shape[0]):
>         s = matches[mt, 1]
>         st = s + m
>         axs[0].plot(np.arange(s, st), c[s : st], c=col)
> 
> axs[1].plot(p[:, 0])
> axs[1].set_ylabel('Matrix profile', fontsize='20')
> plt.show()
> ```

This example doesn't seem to be working for Stumpy 1.11.1.  The `inde` returned from the call to stumpy.motifs is of shape (1, 10) @chrisruk.

**Reply by seanlaw** [reply]
> This example doesn't seem to be working for Stumpy 1.11.1. The inde returned from the call to stumpy.motifs is of shape (1, 10) @chrisruk.

@amurthy-sunysb Please see [this discussion](https://github.com/TDAmeritrade/stumpy/discussions/794)
